crates/prune/src/pruner.rs

```diff
@@ -196,133 +314,141 @@ impl<DB: Database> Pruner<DB> {
         prune_part: PrunePart,
         to_block: BlockNumber,
     ) -> reth_interfaces::Result<Option<RangeInclusive<TxNumber>>> {
-        let from_block_number = provider
+        let from_tx_number = provider
             .get_prune_checkpoint(prune_part)?
-            // Checkpoint exists, prune from the next block after the highest pruned one
-            .map(|checkpoint| checkpoint.block_number + 1)
+            // Checkpoint exists, prune from the next transaction after the highest pruned one
+            .and_then(|checkpoint| match checkpoint.tx_number {
+                Some(tx_number) => Some(tx_number + 1),
+                _ => {
+                    error!(target: "pruner", %prune_part, ?checkpoint, "Expected transaction number in prune checkpoint, found None");
+                    None
+                },
+            })
             // No checkpoint exists, prune from genesis
             .unwrap_or(0);
 
-        // Get first transaction
-        let from_tx_num =
-            provider.block_body_indices(from_block_number)?.map(|body| body.first_tx_num);
-        // If no block body index is found, the DB is either corrupted or we've already pruned up to
-        // the latest block, so there's no thing to prune now.
-        let Some(from_tx_num) = from_tx_num else { return Ok(None) };
-
-        let to_tx_num = match provider.block_body_indices(to_block)? {
+        let to_tx_number = match provider.block_body_indices(to_block)? {
             Some(body) => body,
             None => return Ok(None),
         }
         .last_tx_num();
 
-        Ok(Some(from_tx_num..=to_tx_num))
+        Ok(Some(from_tx_number..=to_tx_number))
     }
 
-    /// Prune receipts up to the provided block, inclusive.
+    /// Prune receipts up to the provided block, inclusive, respecting the batch size.
     #[instrument(level = "trace", skip(self, provider), target = "pruner")]
     fn prune_receipts(
         &self,
         provider: &DatabaseProviderRW<'_, DB>,
         to_block: BlockNumber,
         prune_mode: PruneMode,
     ) -> PrunerResult {
-        let range = match self.get_next_tx_num_range_from_checkpoint(
+        let tx_range = match self.get_next_tx_num_range_from_checkpoint(
             provider,
             PrunePart::Receipts,
             to_block,
         )? {
             Some(range) => range,
             None => {
                 trace!(target: "pruner", "No receipts to prune");
-                return Ok(())
+                return Ok(true)
             }
         };
-        let total = range.clone().count();
+        let tx_range_end = *tx_range.end();
 
-        provider.prune_table_with_iterator_in_batches::<tables::Receipts>(
-            range,
+        let mut last_pruned_transaction = tx_range_end;
+        let (deleted, done) = provider.prune_table_with_range::<tables::Receipts>(
```

Comment:
wondering if we could have `prune_table_with_range` without passing `batch_sizes.receipts` , and just pass the correct tx range? do we expect to have holes in the middle of the range?

Reply:
It was mainly done to replicate the behavior of changset pruning, and we needed to pass the limit manually in there because:
1. `AccountChangeSet` is a dupsort table, so we will have multiple rows per one block
2. `StorageChangeSet` is a dupsort table too, but also has a composite key `(BlockNumber, Address)`, so we will have even more rows per one block

But we indeed do not expect to have any gaps in the middle of receipts, tx senders or tx lookup entries.

Reply:
nevermind, if iterating over hashes, we will skip